sql: subquery_retry test occasionally fails with one duplicate

As part of the #4472 I wanted to enable the `subquery_retry` parallel test. It usually passes but I have seen occasional failures. Every time it failed, the failure showed exactly one duplicate value:

```
--- FAIL: TestParallel (48.88s)
	logic_test.go:654: partestdata/subquery_retry/final:2: expected:
		"0"  
		but found:
		"1"  
```
This might be the same root cause as #4560 but I am filing a separate issue until we know for sure.

I tried using `make stress` to run the test in a loop and found no failures.

I tried making the test longer by increasing the repeat count in https://github.com/cockroachdb/cockroach/blob/master/sql/partestdata/subquery_retry/txn

With `repeat 500`, I hit the failure 2 times and then did not hit it again after a few more runs. I increased to `repeat 2000` (and set `TESTTIMEOUT=30m`) and it still didn't fail.

To run, the `run txn txn txn` line must be uncommented in https://github.com/cockroachdb/cockroach/blob/master/sql/partestdata/subquery_retry/main

--- a/storage/command_queue.go
+++ b/storage/command_queue.go
@@ -129,12 +129,21 @@ func (cq *CommandQueue) GetWait(readOnly bool, wg *sync.WaitGroup, spans ...roac
 		// the new commands overlapped range. Following this strategy, the other dependencies
 		// will be implicitly enforced, which reduces memory utilization and synchronization
 		// costs.
+		// The exception are existing reads: since reads don't wait for each other, an incoming
+		// write must wait for reads even when they are covered by a "later" read (since that
+		// "later" read won't wait for the earlier read to complete).
+		//
+		// TODO(nvanbeschoten): some (likely minimal) gains are possible: The writer does not
+		// have to blindly wait for all reads; only those which aren't already covered directly
+		// by writes are relevant.
 		sort.Sort(overlapSorter(overlaps))
 		for i := len(overlaps) - 1; i >= 0; i-- {
-			c := overlaps[i]
-			if cq.rg.Add(c.Key.Range) {
-				c := c.Value.(*cmd)
-				c.pending = append(c.pending, wg)
+			keyRange, cmd := overlaps[i].Key.Range, overlaps[i].Value.(*cmd)
+			// If we're a write and the current overlap is a read, we always wait
+			// as discussed above. There's some flexibility about adding to the
+			// RangeGroup in that case (currently we do).
+			if cq.rg.Add(keyRange) || (!readOnly && cmd.readOnly) {
+				cmd.pending = append(cmd.pending, wg)
 				wg.Add(1)
 			}
 		}

--- a/storage/command_queue_test.go
+++ b/storage/command_queue_test.go
@@ -22,6 +22,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/util/caller"
 	"github.com/cockroachdb/cockroach/util/leaktest"
 )
 
@@ -84,6 +85,56 @@ func TestCommandQueue(t *testing.T) {
 	if !testCmdDone(cmdDone, 5*time.Millisecond) {
 		t.Fatal("command should finish with no commands outstanding")
 	}
+}
+
+// TestCommandQueueWriteWaitForNonAdjacentRead tests that the command queue
+// lets a writer wait for a read which is separated from it through another
+// read. Since reads don't wait for reads, there was a bug in which the writer
+// would wind up waiting only for one of the two readers under it.
+func TestCommandQueueWriteWaitForNonAdjacentRead(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	cq := NewCommandQueue()
+	key := roachpb.Key("a")
+	// Add a read-only command.
+	wk1 := add(cq, key, nil, true)
+	// Add another one on top.
+	wk2 := add(cq, key, nil, true)
+
+	// A write should have to wait for **both** reads, not only the second
+	// one.
+	wg := sync.WaitGroup{}
+	getWait(cq, key, nil, false /* !readOnly */, &wg)
+
+	assert := func(blocked bool) {
+		cmdDone := waitForCmd(&wg)
+		d := time.Millisecond
+		if !blocked {
+			d *= 5
+		}
+		f, l, _ := caller.Lookup(1)
+		if testCmdDone(cmdDone, d) {
+			if blocked {
+				t.Fatalf("%s:%d: command should not finish with command outstanding", f, l)
+			}
+		} else if !blocked {
+			t.Fatalf("%s:%d: command should not have been blocked", f, l)
+		}
+	}
+
+	// Certainly blocks now.
+	assert(true)
+
+	// The second read returns, but the first one remains.
+	cq.Remove([]interface{}{wk2})
+
+	// Should still block. This being broken is why this test exists.
+	assert(true)
+
+	// First read returns.
+	cq.Remove([]interface{}{wk1})
+
+	// Now it goes through.
+	assert(false)
 }
 
 func TestCommandQueueNoWaitOnReadOnly(t *testing.T) {

